pakyow-core/lib/pakyow/core/router.rb

```diff
@@ -1,174 +1,883 @@
+  # is called in context of its router. This means that any method defined in a
+  # router is available to be called from within a route. For example:
+  #
+  #   Pakyow::App.router do
+  #     def foo
+  #     end
+  #
+  #     get :foo, "/foo" do
+  #       foo
+  #     end
+  #   end
+  #
+  # Including modules works as expected:
+  #
+  #   module AuthHelpers
+  #     def current_user
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router do
+  #     include AuthHelpers
+  #
+  #     get :foo, "/foo" do
+  #       current_user
+  #     end
+  #   end
+  #
+  # See {App.router} for more details on defining routers.
+  #
+  # = Supported HTTP methods
+  #
+  # - +GET+
+  # - +POST+
+  # - +PUT+
+  # - +PATCH+
+  # - +DELETE+
+  #
+  # See {get}, {post}, {put}, {patch}, and {delete}.
+  #
+  # +HEAD+ requests are handled automatically via {Rack::Head}.
+  #
+  # = Building paths for named routes
+  #
+  # Path building is supported via {Controller#path} and {Controller#path_to}.
+  #
+  # = Reusing logic with hooks
+  #
+  # Methods can be defined and used as hooks for a route. For example:
+  #
+  #   Pakyow::App.router do
+  #     def foo
+  #     end
+  #
+  #     get :foo, "/foo", before: [:foo] do
+  #     end
+  #   end
+  #
+  # Before, after, and around hooks are supported in this way.
+  #
+  # = Extending routers
+  #
+  # Extensions can be defined and used to add shared routes to one or more
+  # routers. See {Routing::Extension}.
+  #
+  # = Other routing features
   #
+  # More advanced route features are available, including groups, namespaces,
+  # and templates. See {group}, {namespace}, and {template}.
+  #
+  # = Router subclasses
+  #
+  # It's possible to work with routers outside of Pakyow's DSL. For example:
+  #
+  #   class FooRouter < Pakyow::Router("/foo", after: [:bar])
+  #     def bar
+  #     end
+  #
+  #     default do
+  #       # available at GET /foo
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router << FooRouter
+  #
+  # = Custom matchers
+  #
+  # Routers and routes can be defined with a matcher, which could be a +Regexp+ or
+  # any custom object that implements +match?+. For example:
+  #
+  #   class CustomMatcher
+  #     def match?(path)
+  #       path == "/custom"
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router CustomMatcher.new do
+  #   end
+  #
+  # Custom matchers can also make data available in +params+ by implementing
+  # +match+ and returning an object that implements +named_captures+.
+  # For example:
+  #
+  #   class CustomMatcher
+  #     def match?(path)
+  #       path == "/custom"
+  #     end
+  #
+  #     def match(path)
+  #       return self if match?(path)
+  #     end
+  #
+  #     def named_captures
+  #       { foo: "bar" }
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router CustomMatcher.new do
+  #   end
+  #
+  # When defined on a Router, custom matchers should also implement +sub+,
```

Comment:
Actually, this entire thing is incorrect. We don't currently support path modification from custom matchers. I decided that there wasn't a great reason to support this for now.